File: easy/js/217. Contains Duplicate.js
// Time Complexity = O(n)
// Space Complexity = O(n)
var containsDuplicate = function(nums) {
   const numSet = new Set();

   for(const num of nums){
    if (numSet.has(num)){
        return true;
    }
    numSet.add(num)
   }
   return "asd";
};

function testContainsDuplicate() {
    // Test cases
    const tests = [
      { input: [1, 2, 3, 4, 5], expected: false, description: "No duplicates" },
      { input: [1, 2, 3, 4, 1], expected: true, description: "One duplicate" },
      { input: [1, 1, 2, 3, 3, 4, 5], expected: true, description: "Multiple duplicates" },
      { input: [], expected: false, description: "Empty array" },
      { input: [1], expected: false, description: "Single element" },
    ];
  
    let numOfTest = 0; // Corrected initialization
  
    // Run test cases
    tests.forEach(({ input, expected, description }) => {
      const result = containsDuplicate(input);
      if (result !== expected) {  // If the test fails, update the counter and description
        numOfTest++;
        console.assert(
          result === expected,
          `\nTest Failed: ${description} | For input ${input}, expected ${expected} but got ${result}`
        );
      }
    });
  
    switch (numOfTest) {
      case 0:
        console.log("All tests passed");
        break;
      default:
        console.log(`${numOfTest} Test(s) failed!`);
    }
  }
  
  // Example implementation to test
  function containsDuplicate(nums) {
    const numSet = new Set();
  
    for (const num of nums) {
      if (numSet.has(num)) {
        return true;
      }
      numSet.add(num);
    }
  
    return false;
  }
  
  // Run the tests
  testContainsDuplicate();
  

File: easy/python/217. Contains Duplicate.py
from typing import List

class Solution:
    def containsDuplicate(self, nums: List[int]) -> bool:
        num_set = set()
        
        for num in nums:
            if num in num_set:
                return True
            num_set.add(num)
            
        return False
    
def test_contains_duplicate():
    # Test cases
    tests = [
        {"input": [1, 2, 3, 4, 5], "expected": False, "description": "No duplicates"},
        {"input": [1, 2, 3, 4, 1], "expected": True, "description": "One duplicate"},
        {"input": [1, 1, 2, 3, 3, 4, 5], "expected": True, "description": "Multiple duplicates"},
        {"input": [], "expected": False, "description": "Empty list"},
        {"input": [1], "expected": False, "description": "Single element"},
    ]

    # Initialize variables for counting failed tests
    num_of_failed_tests = 0

    # Create an instance of the Solution class
    solution = Solution()

    # Run test cases
    for test in tests:
        input_data = test["input"]
        expected = test["expected"]
        description = test["description"]

        # Call the containsDuplicate method
        result = solution.containsDuplicate(input_data)
        if result != expected:
            num_of_failed_tests += 1
            print(f"Test Failed: {description} | For input {input_data}, expected {expected} but got {result}")

    # Print the result summary
    print("Test Ended")
    if num_of_failed_tests == 0:
        print("All tests passed")
    else:
        print(f"{num_of_failed_tests} Test(s) failed!")


# Run the tests
test_contains_duplicate()
